msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt

```diff
@@ -0,0 +1,627 @@
+            methodName = "${TAG}.submitChallenge(callback: SubmitChallengeCallback)"
+        )
+        NativeAuthPublicClientApplication.pcaScope.launch {
+            try {
+                val result = submitChallenge(challenge)
+                callback.onResult(result)
+            } catch (e: MsalException) {
+                Logger.error(TAG, "Exception thrown in submitChallenge", e)
+                callback.onError(e)
+            }
+        }
+    }
+
+    /**
+     * Submits the challenge value to the server; Kotlin coroutines variant.
+     *
+     * <strong><u>Warning: this API is experimental. It may be changed in the future without notice. Do not use in production applications.</u></strong>
+     * @return The result of the submit challenge action.
+     */
+    suspend fun submitChallenge(challenge: String): MFASubmitChallengeResult {
+        LogSession.logMethodCall(
+            tag = TAG,
+            correlationId = correlationId,
+            methodName = "${TAG}.submitChallenge(challenge: String)"
+        )
+
+        Logger.warn(TAG, "Warning: this API is experimental. It may be changed in the future without notice. Do not use in production applications.")
+
+        return withContext(Dispatchers.IO) {
+            try {
+                val params = CommandParametersAdapter.createMFASubmitChallengeCommandParameters(
+                    config,
+                    config.oAuth2TokenCache,
+                    challenge,
+                    correlationId,
+                    continuationToken,
+                    scopes
+                )
+
+                val command = MFASubmitChallengeCommand(
+                    parameters = params,
+                    controller = NativeAuthMsalController(),
+                    publicApiId = PublicApiId.NATIVE_AUTH_MFA_SUBMIT_CHALLENGE
+                )
+
+                val rawCommandResult =
+                    CommandDispatcher.submitSilentReturningFuture(command)
+                        .get()
+
+                return@withContext when (val result =
+                    rawCommandResult.checkAndWrapCommandResultType<MFASubmitChallengeCommandResult>()) {
+                    is SignInCommandResult.Complete -> {
+                        val authenticationResult =
+                            AuthenticationResultAdapter.adapt(result.authenticationResult)
+
+                        SignInResult.Complete(
+                            resultValue = AccountState.createFromAuthenticationResult(
+                                authenticationResult = authenticationResult,
+                                correlationId = result.correlationId,
+                                config = config
+                            )
+                        )
+                    }
+                    is SignInCommandResult.IncorrectCode -> {
+                        MFASubmitChallengeError(
+                            errorType = ErrorTypes.INVALID_CODE,
```

Comment:
Is the error type name "invalid code" or "invalid challenge"? On iOS we used "invalid challenge" because the user submits a challenge and not a code

Reply:
Updated